### PR TITLE
[Fix #3103] Make Style/ExtraSpacing not ignore single line hash literals

### LINKED ISSUE #3103: space inside hash values
on this kind of hash
```
{a:   1, b: 2}
```

running rubocop 0.39 on it gives
```
test.rb:1:1: C: Space inside { missing.
{a:   1, b: 2}
^
test.rb:1:14: C: Space inside } missing.
{a:   1, b: 2}
             ^
test.rb:1:15: C: Final newline missing.
{a:   1, b: 2}
```
but it won't warn about the space between the a symbol and the 1 key

doing it on 
```
{
  a:   1,
  b: 2
}
```

does warn about it
```
test.rb:2:3: C: Align the elements of a hash literal if they span more than one line.
  a:   1,
  ^^^^^^
```

### PR DESCRIPTION
**Pull request by tcdowney**
This should fix issue #3103 where the `Style/ExtraSpacing` cop was ignoring extra whitespace within single line hash literals.

Thanks for taking a look at this PR!! 🌝 

-----------------

Before submitting the PR make sure the following are checked:

* [ ] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [ ] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: lib/rubocop/cop/style/extra_spacing.rb [hunk 1]
```diff
@@ -140,15 +140,17 @@ def unary_plus_non_offense?(range)
...
         def ignored_ranges(ast)
           return [] unless ast
 
           @ignored_ranges ||= on_node(:pair, ast).map do |pair|
+            next if pair.parent.single_line?
+
             key, value = *pair
             key.source_range.end_pos...value.source_range.begin_pos
-          end
+          end.compact
```
Comment (tcdowney) [inline lib/rubocop/cop/style/extra_spacing.rb]: I went with this `compact` approach rather than like do something like
```
on_node(:pair, ast).select { |pair| pair.parent.multiline? }
```

because I didn't want to turn the enumerator into an array.  Let me know if there is a cleaner way to do this! :)

File: spec/rubocop/cop/style/extra_spacing_spec.rb [hunk 2]
```diff
@@ -49,6 +49,28 @@
...
+      inspect_source(cop, 'hash = {a:   1,  b:    2}')
+      expect(cop.offenses.size).to eq(3)
+    end
+
+    it 'does not skip single line hash literals with hash rockets' do
+      source = [
+        'let(:single_line_hash) {',
+        '  {"a"   => "1", "b" => "2"}',
+        '}',
+        'let(:multiline_hash) {',
```
Comment (bbatsov) [inline spec/rubocop/cop/style/extra_spacing_spec.rb]: Is the second hash needed here?

File: spec/rubocop/cop/style/extra_spacing_spec.rb [hunk 3]
```diff
@@ -49,6 +49,28 @@
       expect(cop.offenses).to be_empty
     end
 
+    it 'does not skip single line hash literals with symbol keys' do
```
Comment (bbatsov) [inline spec/rubocop/cop/style/extra_spacing_spec.rb]: single-line

File: spec/rubocop/cop/style/extra_spacing_spec.rb (hunk 2 above)
Comment (tcdowney) [inline spec/rubocop/cop/style/extra_spacing_spec.rb]: Maybe not, I just wanted an example that had both single-line and multi-line hashes in the source.

File: spec/rubocop/cop/style/extra_spacing_spec.rb (hunk 3 above)
Comment (tcdowney) [inline spec/rubocop/cop/style/extra_spacing_spec.rb]: Corrected this here and in the Changelog:
https://github.com/bbatsov/rubocop/pull/3565/commits/e33a26e9b69facd11460688bbb6e58722e55a859

File: spec/rubocop/cop/style/extra_spacing_spec.rb (hunk 2 above)
Comment (bbatsov) [inline spec/rubocop/cop/style/extra_spacing_spec.rb]: The point is that this way it's not apparent which hash produced the offense.

File: spec/rubocop/cop/style/extra_spacing_spec.rb [hunk 4]
```diff
@@ -49,6 +49,28 @@
       expect(cop.offenses).to be_empty
     end
 
+    it 'does not skip single-line hash literals with symbol keys' do
```
Comment (bbatsov) [inline spec/rubocop/cop/style/extra_spacing_spec.rb]: I'd rephrase this to the standard "it registers an offense for...

File: spec/rubocop/cop/style/extra_spacing_spec.rb [hunk 5]
```diff
@@ -49,6 +49,28 @@
...
+      inspect_source(cop, 'hash = {a:   1,  b:    2}')
+      expect(cop.offenses.size).to eq(3)
+    end
+
+    it 'does not skip single-line hash literals with hash rockets' do
```
Comment (bbatsov) [inline spec/rubocop/cop/style/extra_spacing_spec.rb]: Same here.

File: spec/rubocop/cop/style/extra_spacing_spec.rb (hunk 4 above)
Comment (tcdowney) [inline spec/rubocop/cop/style/extra_spacing_spec.rb]: Updated this phrasing: 9ce24e4c6cf494853fdd1ce27568744b22219b1f

File: spec/rubocop/cop/style/extra_spacing_spec.rb (hunk 5 above)
Comment (tcdowney) [inline spec/rubocop/cop/style/extra_spacing_spec.rb]: Updated this phrasing: 9ce24e4c6cf494853fdd1ce27568744b22219b1f

File: spec/rubocop/cop/style/extra_spacing_spec.rb (hunk 2 above)
Comment (tcdowney) [inline spec/rubocop/cop/style/extra_spacing_spec.rb]: Good point, I removed the second hash because it wasn't necessary.  I did leave the wrapping `let` block since I wanted the single-line hash to be within a multiline ast.

Updated: 9ce24e4c6cf494853fdd1ce27568744b22219b1f

**Comment by tcdowney** [comment]
@bbatsov I've been rebasing against master as needed and squashing commits as I've been going in this PR.  Let me know if you would prefer for that to happen at the end.
